rllib/algorithms/ppo/utils/env_runner.py

```diff
@@ -0,0 +1,471 @@
+                    fwd_out = self.module[DEFAULT_POLICY_ID].forward_exploration(batch)
+                    # `self.module` is a MARL module.
+                    action_dist_cls = foreach_module(
+                        self.module,
+                        lambda m, mid: (mid, m.get_exploration_action_dist_cls()),
+                    )
+                else:
+                    fwd_out = self.module[DEFAULT_POLICY_ID].forward_inference(batch)
+                    # `self.module` is a MARL module.
+                    action_dist_cls = foreach_module(
+                        self.module,
+                        lambda m, mid: (mid, m.get_inference_action_dist_cls()),
+                    )
+
+                action_dist_cls = {mid: dist_cls for mid, dist_cls in action_dist_cls}
+                action_dist = action_dist_cls[DEFAULT_POLICY_ID].from_logits(
+                    fwd_out[SampleBatch.ACTION_DIST_INPUTS]
+                )
+
+                actions = action_dist.sample().numpy()
+                if STATE_OUT in fwd_out:
+                    states = tree.map_structure(lambda s: s.numpy(), fwd_out[STATE_OUT])
+
+            obs, rewards, terminateds, truncateds, infos = self.env.step(actions)
+            ts += self.num_envs
+            # pbar.update(self.num_envs)
+
+            for i in range(self.num_envs):
+                # Extract state for vector sub_env.
+                s = {k: s[i] for k, s in states.items()}
+                # The last entry in self.observations[i] is already the reset
+                # obs of the new episode.
+                # TODO (simon): This might be unfortunate if a user needs to set a
+                # certain env parameter during different episodes (for example for
+                # benchmarking).
+                if terminateds[i] or truncateds[i]:
+                    # Finish the episode with the actual terminal observation stored in
+                    # the info dict.
+                    self._episodes[i].add_timestep(
+                        # Gym vector env provides the `"final_observation"`.
+                        infos["final_observation"][i],
+                        actions[i],
+                        rewards[i],
+                        state=s,
+                        is_terminated=terminateds[i],
+                        is_truncated=truncateds[i],
+                    )
+                    # Reset h-states to nthe model's intiial ones b/c we are starting a
+                    # new episode.
+                    for k, v in (
+                        self.module[DEFAULT_POLICY_ID].get_initial_state().items()
+                    ):
+                        states[k][i] = v.numpy()
+
+                    done_episodes_to_return.append(self._episodes[i])
+                    # Create a new episode object.
+                    self._episodes[i] = Episode(observations=[obs[i]], states=s)
+                else:
+                    self._episodes[i].add_timestep(
```

Comment:
Many algorithms need extra fetches in the batch. `PPO` needs for example the `ACTION_DIST_INPUTS`, `VF_PREDS`, `ACTION_LOGP` . Before this had been done via the policy, but without it, the `_Episode` might have to take care of it: 
1. using a extra_fetches (maybe `dict`) attribute that can contain several keys and is flexible for the different algorithm's needs,
2. letting each algorithm also override `_Episode` to add the extra fetches as extra attributes to the class as well as implementing the methods to deal with them (I somehow prefer this approach as it is most flexible and also keeps responsibility for sampling for new algorithms mainly at user's side - they decide what they need and how to pack it).

@sven1977 @ArturNiederfahrenhorst Any opinions?

Reply:
Answered in DM on slack :)

Reply:
Hmm, I feel like option 1 is better. I don't think we should require users to custom write their own Episode subclasses for each Algorithm.